include/graaflib/algorithm/cycle_detection.tpp

```diff
@@ -0,0 +1,102 @@
+namespace detail {
+
+template <typename V, typename E, graph_spec S>
+bool do_dfs_directed(const graph<V, E, S>& graph,
+                     std::unordered_map<vertex_id_t, vertex_color>& colored_vertices,
+                     vertex_id_t current) {
+  colored_vertices[current] = vertex_color::VISITED;
+
+  for (const auto& neighbour_vertex : graph.get_neighbors(current)) {
+    if (colored_vertices[neighbour_vertex] == vertex_color::UNVISITED) {
+        if (do_dfs_directed(graph, colored_vertices, neighbour_vertex))
+            return true;
+    } 
+    else if (colored_vertices[neighbour_vertex] == vertex_color::VISITED) {
+      return true;
+    }
+  }
+
+  colored_vertices[current] = vertex_color::NO_CYCLE;
+  return false;
+}
+
+template<typename V, typename E, graph_spec S>
+bool do_dfs_undirected(const graph<V,E,S> & graph, 
+    std::unordered_map<vertex_id_t, bool> & visited_vertices,
+    std::unordered_map<vertex_id_t, vertex_id_t> & parent_vertices,
+    vertex_id_t parent_vertex, vertex_id_t current) {
+
+    visited_vertices[current] = true;
+
+    for (const auto& neighbour_vertex : graph.get_neighbors(current)) {
+        if (neighbour_vertex == parent_vertex) 
+            continue;
+
+        if (visited_vertices[neighbour_vertex]) 
+            return true;
+
+        parent_vertices[neighbour_vertex] = parent_vertex;
+
+        if (do_dfs_undirected(graph, visited_vertices, parent_vertices, neighbour_vertex,
+                            parent_vertices[neighbour_vertex])) {
+                                     return true;
+        }
+    }
+
+    return false;
+}
+
+}  // namespace detail
+
+template <typename V, typename E, graph_spec S>
+bool has_cycle(const graph<V, E, S>& graph) {
+
+  if (graph.is_directed()) {
+        std::unordered_map<vertex_id_t, vertex_color> colored_vertices{};
+
+        for (const auto& vertex : graph.get_vertices()) {
+            if (colored_vertices[vertex.first] == UNVISITED &&
+                detail::do_dfs_directed(graph, colored_vertices, vertex.first)) {
+                        return true;
+            }
+        }
+
+    return false;
+  }
+
+  if (graph.is_undirected()) {
+    // Number of vertices cannot be zero (in case if graph is empty)
+    if (graph.edge_count() >= graph.vertex_count() && graph.vertex_count() > 0) {
+         return true;
+    }
+
+        std::unordered_map<vertex_id_t, bool> visited_vertices{};
+        std::unordered_map<vertex_id_t, vertex_id_t> parent_vertices{};
+
+        for (const auto& vertex : graph.get_vertices()) {
+            if (!visited_vertices.contains(vertex.first) &&
+                 detail::do_dfs_undirected(graph, visited_vertices, parent_vertices,
+                                      vertex.first, parent_vertices[vertex.first])) {
+                         return true;
+            }
+        }
+  }
+
+  return false;
+
+  // We should never reach this
+  std::abort();
+  return false;
```

Comment:
I still think we can simplify this a bit, as we could never reach any line after the first return statement:
```suggestion

  return false;
```